src/ehttpc.erl

```diff
@@ -465,17 +465,67 @@ gun_opts([_ | Opts], Acc) ->
     gun_opts(Opts, Acc).
 
 do_request(Client, head, {Path, Headers}) ->
-    gun:head(Client, Path, Headers);
+    RequestRef = gun:head(Client, Path, Headers),
+    finish_body_call_if_needed(Client, RequestRef, Headers, <<>>),
+    RequestRef;
 do_request(Client, head, Path) ->
     do_request(Client, head, {Path, []});
 do_request(Client, get, {Path, Headers}) ->
-    gun:get(Client, Path, Headers);
+    RequestRef = gun:get(Client, Path, Headers),
+    finish_body_call_if_needed(Client, RequestRef, Headers, <<>>),
+    RequestRef;
 do_request(Client, post, {Path, Headers, Body}) ->
-    gun:post(Client, Path, Headers, Body);
+    RequestRef = gun:post(Client, Path, Headers, Body),
+    finish_body_call_if_needed(Client, RequestRef, Headers, Body),
+    RequestRef;
 do_request(Client, put, {Path, Headers, Body}) ->
-    gun:put(Client, Path, Headers, Body);
+    RequestRef = gun:put(Client, Path, Headers, Body),
+    finish_body_call_if_needed(Client, RequestRef, Headers, Body),
+    RequestRef;
 do_request(Client, delete, {Path, Headers}) ->
-    gun:delete(Client, Path, Headers).
+    RequestRef = gun:delete(Client, Path, Headers),
+    finish_body_call_if_needed(Client, RequestRef, Headers, <<>>),
+    RequestRef.
+
+%% Finish the request only if the headers are set so that gun expect body data
+%% to come with calls to gun:data/4. Otherwise, subsequent request will fail
+%% with a function clause error.
+finish_body_call_if_needed(Client, RequestRef, Headers, Body) ->
+    case is_finish_body_call_needed(Headers, Body) of
+        true ->
+            gun:data(Client, RequestRef, fin, <<>>);
+        false ->
+            ok
+    end.
+
+%% The following function corresponds to request_io_from_headers from
+%% src/gun_http.erl https://github.com/ninenines/gun commit id
+%% 47ec03dcf0346ad827e5c8aa8c2bf9ac35398afe
+%%
+%% Gun checks the headers (in the same way as the following function) to detect
+%% if more data is expected to be provided after the initial request with
+%% gun:data/4.
+is_finish_body_call_needed(Headers, <<>>) ->
+    case lists:keyfind(<<"content-length">>, 1, Headers) of
```

Comment:
just question: do we need to handle capitalization?

Reply:
I think gun normalizes headers to be all lowercase, but that's a great question to check.

Reply:
I cannot find that gun does any normalization of the headers and gun checks if it should wait for more data using lowercase header key as in the check above so I think the check is consistent with how gun works at the moment. So I think this can be merged now if you don't see anything that is wrong?

Reply:
this is from gun's doc
> Header names are case insensitive. Gun converts all the header names
to lowercase, and expects your application to provide lowercase header
names.